tests/integration/models_containers_test.py

```diff
@@ -279,6 +279,28 @@ def test_stats(self):
                     'memory_stats', 'blkio_stats']:
             assert key in stats
 
+    def test_ports(self):
+        client = docker.from_env(version=TEST_API_VERSION)
+        target_ports = {'2222/tcp': None}
+        container = client.containers.run(
+            "alpine", "sleep 100", detach=True,
+            ports=target_ports
+        )
+        self.tmp_containers.append(container.id)
+        container.reload()  # required to get auto-assigned ports
+        actual_ports = container.ports
+        assert sorted(target_ports.keys()) == sorted(actual_ports.keys())
+        for target_client, target_host in target_ports.items():
+            for actual_port in actual_ports[target_client]:
+                actual_keys = sorted(actual_port.keys())
+                assert sorted(['HostIp', 'HostPort']) == actual_keys
+                if target_host is None:
+                    int(actual_port['HostPort'])
+                elif isinstance(target_host, (list, tuple)):
+                    raise NotImplementedError()
```

Comment:
`TypeError` or `ValueError` would be better

Reply:
I don't think so. This is just me being forward-looking. If someone changed the test in the future to have multiple ports, then it would go down this branch. But for the moment it should just report that this has not been implemented (hopefully prompting that person to expand it).

Reply:
[NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError)

    In user defined base classes, abstract methods should raise this exception when they require 
    derived classes to override the method, or while the class is being developed to indicate that 
    the real implementation still needs to be added.

Reply:
Since you are in a test and you have the input and the expected output, IMHO looks more reasonable to assert your expected path like:
```python
assert target_host is None
assert int(actual_port['HostPort']) > 0
```
For the other types, you can can add other tests that does essentially the same but with different structures, like `list` and `tuple`.